gcp_variant_transforms/testing/integration/run_tests_common.py

```diff
@@ -0,0 +1,172 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Common functions and classes that are used by integration tests.
+
+It provides common functions and classes for both run_vcf_to_bq_tests
+(integration test script for vcf_to_bq pipeline) and run_preprocessor_tests
+(integration test script for vcf_to_bq preprocessor pipeline).
+"""
+
+from abc import abstractmethod
+from typing import Dict, List  # pylint: disable=unused-import
+import argparse  # pylint: disable=unused-import
+import json
+import os
+import time
+
+from googleapiclient import discovery
+
+DEFAULT_IMAGE_NAME = 'gcr.io/gcp-variant-transforms/gcp-variant-transforms'
+
+
+class TestCaseFailure(Exception):
+  """Exception for failed test cases."""
+  pass
+
+
+class TestCase(object):
+  """Test case that holds information to run in pipelines API."""
+
+  def __init__(self, test_name, project):
+    # type: (str, str) -> None
+    self._name = test_name
+    self._project = project
+
+  def get_name(self):
+    return self._name
+
+  def run(self, context):
+    service = discovery.build(
+        'genomics', 'v1alpha2', credentials=context.credentials)
+    # The following pylint hint is needed because `pipelines` is a method that
+    # is dynamically added to the returned `service` object above. See
+    # `googleapiclient.discovery.Resource._set_service_methods`.
+    # pylint: disable=no-member
+    request = service.pipelines().run(body=self._pipelines_api_request)
+
+    operation_name = request.execute()['name']
+    response = self._wait_for_operation_done(service, operation_name)
+    self._handle_failure(response)
+
+  def _wait_for_operation_done(self, service, operation_name):
+    """Waits until the operation ``operation_name`` of ``service`` is done."""
+    time.sleep(60)
+    operations = service.operations()
+    request = operations.get(name=operation_name)
+    response = request.execute()
+    while not response['done']:
+      time.sleep(10)
+      response = request.execute()
+    return response
+
+  def _handle_failure(self, response):
+    """Raises errors if test case failed."""
+    if 'error' in response:
+      if 'message' in response['error']:
+        raise TestCaseFailure(response['error']['message'])
+      else:
+        # This case should never happen.
+        raise TestCaseFailure(
+            'No traceback. See logs for more information on error.')
+
+  @abstractmethod
+  def validate_result(self):
```

Comment:
First a general comment that [composition is preferred over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance).

Now, there are certainly cases that inheritance is the right approach but what is the benefit of inheritance in this case? It seems to me that you can achieve what you are doing by passing an instance of `TestCase` to `VcfToBQTestCase` and `PreprocessorTestCase`, right?

Reply:
I like your new design more than the old one; thank you.
But I think you do need to keep some state in the common part of the tests and just relying on direct function calls is not a great idea. Two concrete examples that I can think of are (i) `service` object and other common objects that can be created once and reused between tests and (ii) if we want to move away from the `multiprocess` pattern, it probably makes more sense to keep track of job ids (the ones submitted through Piplines API) in a common class used by both integration tests.

Reply:
Excellent; thanks for making this change. Your `TestRunner` pattern is much cleaner than the old `multiprocess` based pattern.